File: benchmarks/pandas/bench_dataframe_expanding_median_apply.py
"""Benchmark: DataFrame.expanding().median() and .apply(fn) on 10k-row DataFrame."""
import pandas as pd
import numpy as np
import json
import time

ROWS = 10_000
WARMUP = 2
ITERATIONS = 5

a = np.sin(np.arange(ROWS) * 0.05) * 100
b = np.cos(np.arange(ROWS) * 0.05) * 80
df = pd.DataFrame({"a": a, "b": b})

sum_fn = lambda x: x.sum()

for _ in range(WARMUP):
    df.expanding().median()
    df.expanding().apply(sum_fn, raw=True)

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.expanding().median()
    df.expanding().apply(sum_fn, raw=True)
total = time.perf_counter() - start

print(json.dumps({
    "function": "dataframe_expanding_median_apply",
    "mean_ms": total / ITERATIONS * 1000,
    "iterations": ITERATIONS,
    "total_ms": total * 1000,
}))

File: benchmarks/tsb/bench_dataframe_expanding_median_apply.ts
/**
 * Benchmark: DataFrameExpanding.median() and .apply(fn) on 10k-row DataFrame.
 * Outputs JSON: {"function": "dataframe_expanding_median_apply", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame } from "../../src/index.js";

const ROWS = 10_000;
const WARMUP = 2;
const ITERATIONS = 5;

const a = Array.from({ length: ROWS }, (_, i) => Math.sin(i * 0.05) * 100);
const b = Array.from({ length: ROWS }, (_, i) => Math.cos(i * 0.05) * 80);
const df = DataFrame.fromColumns({ a, b });

const sumFn = (vals: readonly number[]) => vals.reduce((acc, v) => acc + v, 0);

for (let i = 0; i < WARMUP; i++) {
  df.expanding().median();
  df.expanding().apply(sumFn);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  df.expanding().median();
  df.expanding().apply(sumFn);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dataframe_expanding_median_apply",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
